src/Plugins/DBFTPlugin/Consensus/ConsensusService.cs

```diff
@@ -85,29 +90,30 @@ private void InitializeConsensus(byte viewNumber)
                 Log($"View changed: view={viewNumber} primary={context.Validators[context.GetPrimaryIndex((byte)(viewNumber - 1u))]}", LogLevel.Warning);
             Log($"Initialize: height={context.Block.Index} view={viewNumber} index={context.MyIndex} role={(context.IsPrimary ? "Primary" : context.WatchOnly ? "WatchOnly" : "Backup")}");
             if (context.WatchOnly) return;
+            blockGenTime = neoSystem.GetBlockGenTime();
             if (context.IsPrimary)
             {
                 if (isRecovering)
                 {
-                    ChangeTimer(TimeSpan.FromMilliseconds(neoSystem.Settings.MillisecondsPerBlock << (viewNumber + 1)));
+                    ChangeTimer(TimeSpan.FromMilliseconds((int)blockGenTime.TotalMilliseconds << (viewNumber + 1)));
                 }
                 else
                 {
-                    TimeSpan span = neoSystem.Settings.TimePerBlock;
-                    if (block_received_index + 1 == context.Block.Index)
+                    if (block_received_index + 1 == context.Block.Index && onPrepareBlockIndex + 1 == context.Block.Index)
                     {
-                        var diff = TimeProvider.Current.UtcNow - block_received_time;
-                        if (diff >= span)
-                            span = TimeSpan.Zero;
+                        // Include the consensus time into the consensus intervals.
+                        var diff = TimeProvider.Current.UtcNow - onPrepareReceivedTime;
```

Comment:
Ditto, we need a separate PR for these changes.

Reply:
This does not matter, has being evaluated in real testnet and passed stress test, this pr has being there for too long,  will have another solid test and run, and most importantly this is not a protocol levele change, so please allow me to keep it.

Reply:
It matters since it's a separate logic change that affects the consensus behaviour and that is not related to 3 seconds consensus in any way. It should be reviewed carefully and we'd better so it in a separate PR.

@roman-khimov

Reply:
> It matters since it's a separate logic change that affects the consensus behaviour and that is not related to 3 seconds consensus in any way. It should be reviewed carefully and we'd better so it in a separate PR.
> 
> @roman-khimov

That is realated, 3 seconds will cause higher delay, this change is for solving that issue and make the consensus more stable and mitigate the delay impact.

Reply:
This particular change could've been merged months ago, it's pretty clear on its own and tested well enough now. But for whatever reason it's sitting here with a number of Policy changes that we're trying to settle. An option too, just some overhead for everyone involved.

Reply:
Two prs is better, easier to review and don't miss anything